toolchain/lc-compile/src/emit.cpp

```diff
@@ -1719,3 +1729,253 @@ void OutputEnd(void)
     
     fclose(s_output);
 }
+
+//////////
+
+struct DependMapping
+{
+    NameRef module;
+    const char *source;
+    char *interface;
+    time_t source_time;
+    time_t interface_time;
+    bool is_interface;
+    bool processed;
+    bool changed;
+};
+
+struct DependDependency
+{
+    NameRef module;
+    NameRef dependency;
+};
+
+int DependencyMode = 0;
+
+static DependMapping *s_depend_mappings;
+static int s_depend_mapping_count;
+
+static DependDependency *s_depend_deps;
+static int s_depend_dep_count;
+
+static bool DependFindMapping(NameRef p_module, DependMapping **r_mapping)
+{
+    for(int i = 0; i < s_depend_mapping_count; i++)
+        if (p_module == s_depend_mappings[i] . module)
+        {
+            if (r_mapping != NULL)
+                *r_mapping = &s_depend_mappings[i];
+            return true;
+        }
+    
+    return false;
+}
+
+static bool DependProcess(NameRef p_module)
+{
+    DependMapping *t_mapping;
+    if (!DependFindMapping(p_module, &t_mapping))
+        return false;
+    
+    const char *t_module_name;
+    GetStringOfNameLiteral(p_module, &t_module_name);
+    
+    if (t_mapping -> processed)
+        return t_mapping -> changed;
+    
+    // Mark this module as being processed.
+    t_mapping -> processed = true;
+    
+    // If this module is an interface, then we can't do anything with the info
+    // we've got so skip.
+    if (t_mapping -> is_interface)
+        return false;
+    
+    // If our source is more recent than our interface then we must recompile.
+    bool t_changed;
+    t_changed = t_mapping -> source_time > t_mapping -> interface_time;
+    if (t_changed)
+        Debug_Depend("Recompiling '%s' as source newer than interface", t_module_name);
+    
+    // If any dependent modules need recompiling, we must recompile after.
+    for(int i = 0; i < s_depend_dep_count; i++)
+    {
+        if (s_depend_deps[i] . module == p_module)
+        {
+            // If a dependency has changed, then we must be recompiled.
+            if (DependProcess(s_depend_deps[i] . dependency))
+            {
+                const char *t_depend_name;
+                GetStringOfNameLiteral(s_depend_deps[i] . dependency, &t_depend_name);
+                Debug_Depend("Recompiling '%s' as dependency '%s' changed", t_module_name, t_depend_name);
+                t_changed = true;
+            }
+        }
+    }
+    
+    // If any dependent module interfaces are more recent than our interface then
+    // we must recompile.
+    for(int i = 0; i < s_depend_dep_count; i++)
```

Comment:
Yes - I think so - only the first loop has to run as it processes the dependencies to produce the ordering.